Add spilling to HashAgrregation

oerling opened the pull request.

mbasmanova commented:
@oerling Orri, ci/circleci: macos-build is failing. Would you take a look?

oerling commented:
Added the explanation.

oerling commented:
Thank you, Masha.  I think this passes now.

 

 

From: Maria Basmanova ***@***.***> 
Sent: Thursday, April 28, 2022 5:19 AM
To: facebookincubator/velox ***@***.***>
Cc: oerling ***@***.***>; Mention ***@***.***>
Subject: Re: [facebookincubator/velox] Add spilling to HashAgrregation (PR #1447)

 

@oerling <https://github.com/oerling>  Orri, ci/circleci: macos-build is failing. Would you take a look?

—
Reply to this email directly, view it on GitHub <https://github.com/facebookincubator/velox/pull/1447#issuecomment-1112134368> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPTYZBR7KI2B24UQDK7TVHJ63ZANCNFSM5TZJODPQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>

aditi-pandit commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
+    return;
+  }
+
+  // Check if can inscrease reservation. The The increment is the larger of
```
Question : What is the intuition for using 1/4 of the current reservation ?

mbasmanova replied:
typo: "The The"

mbasmanova commented on velox/core/QueryCtx.h:
```diff
@@ -207,8 +220,8 @@ class ExecCtx : public Context {
   // Pool for all Buffers for this thread
-  memory::MemoryPool* pool_;
-  QueryCtx* queryCtx_;
+  memory::MemoryPool* FOLLY_NONNULL pool_;
```
It would be nice to extract adding FOLLY_NONNULL to existing code into a separate PR. This would make this PR shorter and allow reviewers to focus on material changes.

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -613,5 +678,88 @@ TEST_F(AggregationTest, partialAggregationMemoryLimit) {
   assertQuery(params, "SELECT c0, count(1) FROM tmp GROUP BY 1");
 }
 
+TEST_F(AggregationTest, spill) {
```
A single test for all of spilling logic feels not enough. Do you think it covers all the code paths?

mbasmanova commented on velox/exec/RowContainer.h:
```diff
@@ -320,6 +320,10 @@ class RowContainer {
   }
 
+  // Returns a cap on  extra memory that may be needed when adding 'numRows'
+  // with 'bytes in in flat vector space consumption.
```
typo: "in in"

What does "bytes in in flat vector space consumption" mean?

mbasmanova commented on velox/exec/RowContainer.cpp:
```diff
@@ -531,4 +531,15 @@ void RowContainer::setProbedFlag(char** rows, int32_t numRows) {
+      ? std::min<int64_t>(0, stringAllocator_.freeSpace() - flatBytes)
+      : 0;
+  return bits::roundUp(needRows * fixedRowSize_, kAllocUnit) +
+      -+bits::roundUp(needBytes, kAllocUnit);
```
What is "-+" here for?

mbasmanova commented on velox/exec/Operator.h:
```diff
@@ -107,6 +107,7 @@ struct OperatorStats {
   CpuWallTiming finishTiming;
 
   MemoryStats memoryStats;
+  uint64_t spilledBytes{0};
```
Would you add a small comment? Would it make sense to also track number of spilled rows?

mbasmanova commented on velox/exec/LocalPlanner.cpp:
```diff
@@ -83,6 +83,12 @@ OperatorSupplier makeConsumerSupplier(
     };
+  } else if (
+      auto join =
+          std::dynamic_pointer_cast<const core::HashJoinNode>(planNode)) {
```
What is this change for? It is surprising to see a join logic in a diff that focuses on aggregation.

mbasmanova commented on velox/exec/HashTable.h:
```diff
@@ -143,6 +143,15 @@ class BaseHashTable {
+  // Returns the size  in bytes for hash table structures given the count of
+  // distinct keys.
+  static uint64_t tableByteSize(int32_t numDistinct) {
+    return bits::nextPowerOfTwo(numDistinct) * 9;
```
Would you document the constant 9?

mbasmanova commented on velox/exec/HashAggregation.h:
```diff
@@ -70,6 +70,9 @@ class HashAggregation : public Operator {
   bool pushdownChecked_ = false;
   bool mayPushdown_ = false;
+
+  // Intermediate types of aggregates. Used for spilling
```
Period is missing at the end of the sentence.

mbasmanova commented on velox/exec/Aggregate.h:
```diff
@@ -193,6 +202,13 @@ class Aggregate {
+  // Returns the type that extractAccumulator will produce. This is
+  // different from resultType_ for complex final aggregations. Needed
+  // for e.g. spilling.
+  virtual TypePtr accumulatorType() const {
```
This API seems redundant. We already know signatures of all the aggregate functions and can use SignatureBinder to compute intermediate and final types given raw input types. See resolveAggregateType in PlanBuilder.cpp.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -35,6 +35,17 @@ bool areAllLazyNotLoaded(const std::vector<VectorPtr>& vectors) {
   });
 }
+
+std::string makeSpillPath(bool isPartial, const OperatorCtx& operatorCtx) {
```
Consider using std::optional for return type for readability. Same for spillPath in config.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -35,6 +35,17 @@ bool areAllLazyNotLoaded(const std::vector<VectorPtr>& vectors) {
+  }
+  auto path = operatorCtx.task()->queryCtx()->config().spillPath();
+  if (!path.empty()) {
+    return path + "/" + operatorCtx.task()->taskId();
```
Would it make sense to make separate directory for each operator?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -356,11 +379,13 @@ const SelectivityVector& GroupingSet::getSelectivityVector(
     int32_t batchSize,
-    bool isPartial,
-    RowContainerIterator* iterator,
+    RowContainerIterator* FOLLY_NONNULL iterator,
```
getGlobalAggregationOutput was modified to take iterator by reference; let's make the same change here for consistency

mbasmanova reviewed (changes requested):
@oerling I skipped the PR and have some initial comments. I haven't read velox/exec/GroupingSet.cpp changes yet.

mbasmanova commented on velox/exec/Aggregate.h:
```diff
@@ -65,6 +65,15 @@ class Aggregate {
+  // Returns the size for a single-value variable length
+  // accumulator. For example, map_agg has 4xthe initial array size as
+  // overhead even for a single key-value pair. This is on top of
+  // accumulatorFixedWidthBytes() and the serialized size of the
```
typo: accumulatorFixedWidthBytes -> accumulatorFixedWidthSize

The description of the method is not very clear. What is "the size for a single-value variable length accumulator". Is this an estimate of the size of the variable-width part of the accumulator? Are there any suggestions for function authors on how to calculate these estimates?

It would be more robust for default implementation to throw if isFixedSize is false. Otherwise, it is easy to forget to implement this method.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
+} // namespace
+
+void GroupingSet::ensureInputFits(const RowVectorPtr& input) {
+  if (isPartial_ || spillPath_.empty()) {
```
Perhaps, add a small comment to explain that spilling is enabled only for final aggregations.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
+  if (!numDistinct) {
+    return;
+  }
+  auto tableIncrement = numDistinct + input->size() >= table_->rehashSize()
```
Shouldn't these calculations take into account activeRows_, i.g. look for space only for a subset of rows?

mbasmanova replied:
I wonder if it would be cleaner to move this logic into HashTable and expose it via a HashTable::tableSizeIncrease(newNumDistinct).

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
+          BaseHashTable::tableByteSize(numDistinct)
+      : 0;
+
+  int64_t retainedSize = input->retainedSize();
```
Retained size may be a lot larger or smaller than flattened size. How is it used here?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
+  auto outOfLineBytes =
+      rows->stringAllocator().retainedSize() - outOfLineFreeBytes;
+  auto outOfLineBytesPerRow =
+      (outOfLineBytes / numDistinct) + minVariableWidthAccumulatorBytes_;
```
Instead of relying on individual aggregates to provide reasonable estimates, can we use rows->stringAllocator().retainedSize() / numDistinct as the estimate?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +446,244 @@ const HashLookup& GroupingSet::hashLookup() const {
+        *rows,
+        [&](folly::Range<char**> rows) { table_->erase(rows); },
+        ROW(std::move(names), std::move(types)),
+        HashBitRange(29, 31),
```
Why HashBitRange(29, 31)?

mbasmanova replied:
Why HashBitRange(29, 31)?

mbasmanova reviewed (commented):
@oerling Some further comments.

mbasmanova reviewed (commented):
@oerling I'm about to review this PR, but I noticed it includes changes to third_party/xsimd. I assume these are accidental. Would you update the PR to remove these changes?

mbasmanova commented:
@oerling Orri, I'm seeing that "This branch has conflicts that must be resolved". Would you rebase and update the PR?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {
+  auto candidate = &tracker;
+  while (candidate) {
+    auto limit = candidate->maxTotalBytes();
+    if (limit == memory::kMaxMemory && candidate->parent()) {
```
What does 'limit == memory::kMaxMemory' mean? Unlimited?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {
 }
 
+namespace {
+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {
```
Any particular reason to have this logic inside of the aggregation operator and not as a method on MemoryUsageTracker?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {
+int64_t estimateSerializedSize(const VectorPtr& vector) {
+  vector_size_t bytes = 0;
+  auto bytesPtr = &bytes;
+  VectorStreamGroup::estimateSerializedSize(
```
It would be nice to add a version of estimateSerializedSize that takes a vector and returns the estimate in bytes for the whole vector:

```
auto bytes = VectorStreamGroup::estimateSerializedSize(vector);
```

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {
+  }
+  auto numDistinct = table_->numDistinct();
+  if (!numDistinct) {
+    // Table is empty. Nothing to spill.
```
This method is called 'ensureInputFits'. The face that table_ is empty doesn't necessarily mean that input will fit, does it?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {
+  return false;
+}
+
+int64_t estimateSerializedSize(const VectorPtr& vector) {
```
This method appears unused. If that's the case, let's remove.

mbasmanova reviewed (commented):
@oerling I tried to read through the change to GroupingSet, but it is super hard. I'll try again later, but have some questions based on what I read so far.

oerling commented:
Masha, thank you. Addressed the comments and updated.

 

 

In velox/exec/Aggregate.cpp:

> @@ -77,4 +78,21 @@ std::unique_ptr<Aggregate> Aggregate::create(

   VELOX_USER_FAIL("Aggregate function not registered: {}", name);

Is this the case if we run partially-streaming aggregation?

 

A: Partial and streaming aggregations do not spill.

 

 

 

In velox/exec/GroupingSet.cpp:

> @@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {

   return *lookup_;

+  // increase reservation.

+  auto candidate = &tracker;

+  while (candidate) {

+    auto limit = candidate->maxTotalBytes();

+    if (limit == memory::kMaxMemory && candidate->parent()) {

What does 'limit == memory::kMaxMemory' mean? Unlimited?

 

A: It means no limit.

 

In velox/exec/GroupingSet.cpp:

> @@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {

   return *lookup_;

}

+namespace {

+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {

Any particular reason to have this logic inside of the aggregation operator and not as a method on MemoryUsageTracker?

 

A: This belongs more properly to Task. But to keep things simple and incremental we do not put this in Task yet. Added comment.

 

 

 

+  // spillPath is set.

+  if (isPartial_ || !spillPath_.has_value()) {

+    return;

+  }

+  auto numDistinct = table_->numDistinct();

+  if (!numDistinct) {

+    // Table is empty. Nothing to spill.

This method is called 'ensureInputFits'. The face that table_ is empty doesn't necessarily mean that input will fit, does it?

 

 

A: the space to aggregate a batch is not much larger than the batch itself. Except maybe for a two column batch with a very expensive digest accumulator. But even so, we  are talking some tens of bytes a row. If this does not fit then we are so out of resources we may as well error out.

 

 

In velox/exec/GroupingSet.cpp:

> +  return bytes;

+}

+} // namespace

+

+void GroupingSet::ensureInputFits(const RowVectorPtr& input) {

+  // Spilling is considered if this is a final or single aggregation and

+  if (!numDistinct) {

+    // Table is empty. Nothing to spill.

+    return;

+  }

+  auto tableIncrement = table_->hashTableSizeIncrease(input->size());

Since we are not adding all of input, but only activeRows_, shouldn't this logic use activeRows_.countSelected() instead of input->size()?

 

A: Overestimating a bit is appropriate here.  A large difference between active and size will be uncommon.

 

 

In velox/exec/GroupingSe

+      rows->stringAllocator().retainedSize() - outOfLineFreeBytes;

+  int64_t flatBytes = input->estimateFlatSize();

estimateFlatSize is just an estimate; if this estimate a way lower than actual, we may fail to spill and run out of memory. Is this the case?

 

 

A: This is a soft check for cases that clearly need to spill. If this check does not spill but goes ahead and the op exceeds the cap, this is a hard out of cap that leads to running the cap increase negotiation. This Will evict cache or spill somebody else. 

 If this can't be done then we may as well error out. There is the outside possibility that there is nobody else.  Spill cannot be called on the allocating aggregation because it is not in a defined state, i.e. in the middle of an accumulator. The action is then to raise the cap enough the batch finishes and set a state forcing  a full spill after finishing the batch.

 

The logic for this comes in the memory atrbitration diff. For immediate purposes without outlier accumulator behavior the present is good enough.

 

In velox/exec/Operator.h:

> @@ -107,6 +107,7 @@ struct OperatorStats {

   CpuWallTiming finishTiming;

   MemoryStats memoryStats;

+  uint64_t spilledBytes{0};

Would it make sense to also track number of spilled rows?

 

A: The time is relative to bytes and the Presto reporting structures only hav bytes. I don't think rows would add very much here. The other stats give an idea of row size.

 

+        HashBitRange(29, 31),

Why HashBitRange(29, 31)?

 

A: Bit field of 2 bits somewhere in the hash number. Any field would do. For hash join the field will slide for each recursive level of spilling, so it is not globally constant. In practice, think of 15 drivers in a final agg. The per query cap is, say 7 GB, with 1/2 GB per driver. So at spill time the table is 512MB which divides in 4 hash partitions. There is up to 128MB to spill at a time. Spilling half of each partition would make 64MB spill files,. Unspilling these would be 4 sets of n streams of 64MB. This would be a practical size.

 

Suppose We need to free space and ont just stop growing. Then we would pick one of the 15 drivers and spill all its 512MB  and make this much free space. This is still reasonable size steps. 

 

The principle is good, we will tune after getting some more experience running.

mbasmanova commented:
> The time is relative to bytes and the Presto reporting structures only hav bytes. 

We do not need to limit ourselves by what Presto supports. I think reporting # of spilled rows in addition to # of bytes would be helpful.

mbasmanova commented on velox/core/QueryConfig.h:
```diff
@@ -93,6 +93,8 @@ class QueryConfig {
 
   static constexpr const char* kCreateEmptyFiles = "driver.create_empty_files";
 
+  static constexpr const char* kSpillPath = "experimental.spiller-spill-path";
```
Let's not add "experimental." prefixes to property names. It might seem like a good idea at first, but in practice we never go back and remove these, so these prefixes become confusing.

mbasmanova commented on velox/core/QueryConfig.h:
```diff
@@ -170,14 +172,23 @@ class QueryConfig {
     return get<bool>(kExprEvalSimplified, false);
   }
 
+  std::optional<std::string> spillPath() const {
```
Would you add a comment to explain what this is used for and what should be the format of the path? Do we support different storage adapters here or only local files? What kind of structure do we create within the path? Who is responsible for cleaning up the no longer needed files on crash, etc.?

mbasmanova commented on velox/core/QueryCtx.h:
```diff
@@ -133,8 +138,12 @@ class QueryCtx : public Context {
+
  private:
-  static Config* getEmptyConfig() {
+  static Config* FOLLY_NONNULL getEmptyConfig() {
```
Would you extract FOLLY_NONNULL changes into a separate PR? This PR is super complex and having non-material changes makes reviewing even harder.

mbasmanova commented on velox/exec/GroupingSet.h:
```diff
@@ -62,6 +65,19 @@ class GroupingSet {
 
   const HashLookup& hashLookup() const;
 
+  // Spills content until under 'targetRows' and under 'targetBytes'
```
Use /// for method and class comments.

mbasmanova commented on velox/exec/GroupingSet.h:
```diff
@@ -86,13 +102,49 @@ class GroupingSet {
+  // otherwise.
+  void extractGroups(folly::Range<char**> groups, const RowVectorPtr& result);
+
+  bool getOutputWithSpill(const RowVectorPtr& result);
```
Would you add a comment?

mbasmanova commented on velox/exec/GroupingSet.h:
```diff
@@ -86,13 +102,49 @@ class GroupingSet {
+
+  bool getOutputWithSpill(const RowVectorPtr& result);
+
+  bool mergeNext(const RowVectorPtr& result);
```
Would you add a comment?

mbasmanova commented on velox/exec/GroupingSet.h:
```diff
@@ -86,13 +102,49 @@ class GroupingSet {
+  // state with no data accumulated. This is called each time a new
+  // key is received from a merge of spilled data. After this
+  // updateRow() is called on the same element and on every subsequent
+  // element read from the stream until a new key is seen, when we
```
typo: when -> then

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+// unlikely. Otherwise attempts the reservation increment and
+// returns true if succeeded. This will be moved to Task when adding
+// memory contention arbitration.
+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {
```
I feel that maybeReserve logic deserves a test. Like you mentioned it doesn't belong to GroupingSet to begin this. Consider extracting it into a separate PR.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+        *rows,
+        [&](folly::Range<char**> rows) { table_->erase(rows); },
+        ROW(std::move(names), std::move(types)),
+        // Spill up to 4 partitions based on bits 29 and 30 of the hash number.
```
Would you add a comment explaining that the choice of bits 29 and 30 is arbitrary?

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+    for (auto& hasher : table_->hashers()) {
+      keyTypes.push_back(hasher->type());
+    }
+    mergeRows_ = std::make_unique<RowContainer>(
```
Passing so many unnecessary parameters makes it hard to read this code. Let's introduce a RowContainer's constructor that takes just keyTypes, nullableKeys and aggregates.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+    uint64_t bytes = 0;
+    vector_size_t numGroups = 0;
+    auto limit = std::min<size_t>(
+        1000, nonSpilledRows_.value().size() - nonSpilledIndex_);
```
Please, comment about the choice of 1000.

mbasmanova commented on velox/exec/GroupingSet.cpp:
```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
+
+bool GroupingSet::mergeNext(const RowVectorPtr& result) {
+  constexpr int32_t kBatchBytes = 1 << 20; // 1MB
+  if (mergeSelection_.size() < 64) {
```
Why 1MB and 64? Please, comment to explain. It would be nice to collect all the constants to a single place and document these together.

mbasmanova reviewed (changes requested):
@oerling Orri, thank you for iterating on this PR. Some further comments and a high-level question. What types of workload did you try this with? What does the profile look like? What's expensive? How much CPU time and latency the spill is adding?

oerling commented:
Thank you, Masha. I edited the item and extracted maybeReserve out.

 

The ‘experimental’ name is literally from Presto. It would not have occurred to me to call something experimental unless it was already called that. 

 

The test is dominated by spookyhash of string. This is folly::hash. One could have a faster hash for strings with the crc 32 instruction. But this is not the thing here. The test is 1.09s with spilling and 1s without. Spilling spends time in sort and compare, which is normal since that’s the only non-linear item.  The test is a group by with array_agg.  The profile is plenty good enough. To have real experience running this in situ, need memory arbitration and some larger queries.

 

 

 

 

 

 

 

 

From: Maria Basmanova ***@***.***> 
Sent: Wednesday, May 18, 2022 5:32 AM
To: facebookincubator/velox ***@***.***>
Cc: oerling ***@***.***>; Mention ***@***.***>
Subject: Re: [facebookincubator/velox] Add spilling to HashAgrregation (PR #1447)

 

@mbasmanova requested changes on this pull request.

@oerling <https://github.com/oerling>  Orri, thank you for iterating on this PR. Some further comments and a high-level question. What types of workload did you try this with? What does the profile look like? What's expensive? How much CPU time and latency the spill is adding?

  _____  

In velox/core/QueryConfig.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875819842> :

> @@ -93,6 +93,8 @@ class QueryConfig {
 
   static constexpr const char* kCreateEmptyFiles = "driver.create_empty_files";
 
+  static constexpr const char* kSpillPath = "experimental.spiller-spill-path";

Let's not add "experimental." prefixes to property names. It might seem like a good idea at first, but in practice we never go back and remove these, so these prefixes become confusing.

  _____  

In velox/core/QueryConfig.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875821206> :

> @@ -170,14 +172,23 @@ class QueryConfig {
     return get<bool>(kExprEvalSimplified, false);
   }
 
+  std::optional<std::string> spillPath() const {

Would you add a comment to explain what this is used for and what should be the format of the path? Do we support different storage adapters here or only local files? What kind of structure do we create within the path? Who is responsible for cleaning up the no longer needed files on crash, etc.?

  _____  

In velox/core/QueryCtx.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875821940> :

>   private:
-  static Config* getEmptyConfig() {
+  static Config* FOLLY_NONNULL getEmptyConfig() {

Would you extract FOLLY_NONNULL changes into a separate PR? This PR is super complex and having non-material changes makes reviewing even harder.

  _____  

In velox/exec/GroupingSet.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875822725> :

> @@ -62,6 +65,19 @@ class GroupingSet {
 
   const HashLookup& hashLookup() const;
 
+  // Spills content until under 'targetRows' and under 'targetBytes'

Use /// for method and class comments.

  _____  

In velox/exec/GroupingSet.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875823880> :

> @@ -86,13 +102,49 @@ class GroupingSet {
   // index for this aggregation), otherwise it returns reference to activeRows_.
   const SelectivityVector& getSelectivityVector(size_t aggregateIndex) const;
 
+  // Checks if input will fit in the existing memory and increases
+  // reservation if not. If reservation cannot be increased, spills
+  // enough to make 'input' fit.
+  void ensureInputFits(const RowVectorPtr& input);
+
+  // Copies the grouping keys and aggregates for 'groups' into 'result' If
+  // partial output, extracts the intermediate type for aggregates, final result
+  // otherwise.
+  void extractGroups(folly::Range<char**> groups, const RowVectorPtr& result);
+
+  bool getOutputWithSpill(const RowVectorPtr& result);

Would you add a comment?

  _____  

In velox/exec/GroupingSet.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875824033> :

> @@ -86,13 +102,49 @@ class GroupingSet {
   // index for this aggregation), otherwise it returns reference to activeRows_.
   const SelectivityVector& getSelectivityVector(size_t aggregateIndex) const;
 
+  // Checks if input will fit in the existing memory and increases
+  // reservation if not. If reservation cannot be increased, spills
+  // enough to make 'input' fit.
+  void ensureInputFits(const RowVectorPtr& input);
+
+  // Copies the grouping keys and aggregates for 'groups' into 'result' If
+  // partial output, extracts the intermediate type for aggregates, final result
+  // otherwise.
+  void extractGroups(folly::Range<char**> groups, const RowVectorPtr& result);
+
+  bool getOutputWithSpill(const RowVectorPtr& result);
+
+  bool mergeNext(const RowVectorPtr& result);

Would you add a comment?

  _____  

In velox/exec/GroupingSet.h <https://github.com/facebookincubator/velox/pull/1447#discussion_r875824496> :

> +
+  // Copies the grouping keys and aggregates for 'groups' into 'result' If
+  // partial output, extracts the intermediate type for aggregates, final result
+  // otherwise.
+  void extractGroups(folly::Range<char**> groups, const RowVectorPtr& result);
+
+  bool getOutputWithSpill(const RowVectorPtr& result);
+
+  bool mergeNext(const RowVectorPtr& result);
+
+  // Initializes a new row in 'mergeRows' with the keys from the
+  // current element from 'keys'. Accumulators are left in the initial
+  // state with no data accumulated. This is called each time a new
+  // key is received from a merge of spilled data. After this
+  // updateRow() is called on the same element and on every subsequent
+  // element read from the stream until a new key is seen, when we

typo: when -> then

  _____  

In velox/exec/GroupingSet.cpp <https://github.com/facebookincubator/velox/pull/1447#discussion_r875827713> :

> @@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
   return *lookup_;
 }
 
+namespace {
+// Checks if it is likely that the reservation on 'tracker' can be
+// incremented by 'increment'. Returns false if this seems
+// unlikely. Otherwise attempts the reservation increment and
+// returns true if succeeded. This will be moved to Task when adding
+// memory contention arbitration.
+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {

I feel that maybeReserve logic deserves a test. Like you mentioned it doesn't belong to GroupingSet to begin this. Consider extracting it into a separate PR.

  _____  

In velox/exec/GroupingSet.cpp <https://github.com/facebookincubator/velox/pull/1447#discussion_r875829370> :

> +void GroupingSet::spill(int64_t targetRows, int64_t targetBytes) {
+  if (!spiller_) {
+    auto rows = table_->rows();
+    auto types = rows->keyTypes();
+    types.insert(
+        types.end(), intermediateTypes_.begin(), intermediateTypes_.end());
+    std::vector<std::string> names;
+    for (auto i = 0; i < types.size(); ++i) {
+      names.push_back(fmt::format("s{}", i));
+    }
+    auto fileSize = mappedMemory_->tracker()->getCurrentUserBytes() / 4;
+    spiller_ = std::make_unique<Spiller>(
+        *rows,
+        [&](folly::Range<char**> rows) { table_->erase(rows); },
+        ROW(std::move(names), std::move(types)),
+        // Spill up to 4 partitions based on bits 29 and 30 of the hash number.

Would you add a comment explaining that the choice of bits 29 and 30 is arbitrary?

  _____  

In velox/exec/GroupingSet.cpp <https://github.com/facebookincubator/velox/pull/1447#discussion_r875831595> :

> +        spillPath_.value(),
+        fileSize,
+        Spiller::spillPool(),
+        spillExecutor_);
+  }
+  spiller_->spill(targetRows, targetBytes, spillIterator_);
+}
+
+bool GroupingSet::getOutputWithSpill(const RowVectorPtr& result) {
+  if (outputPartition_ == -1) {
+    mergeArgs_.resize(1);
+    std::vector<TypePtr> keyTypes;
+    for (auto& hasher : table_->hashers()) {
+      keyTypes.push_back(hasher->type());
+    }
+    mergeRows_ = std::make_unique<RowContainer>(

Passing so many unnecessary parameters makes it hard to read this code. Let's introduce a RowContainer's constructor that takes just keyTypes, nullableKeys and aggregates.

  _____  

In velox/exec/GroupingSet.cpp <https://github.com/facebookincubator/velox/pull/1447#discussion_r875832228> :

> +        false,
+        mappedMemory_,
+        ContainerRowSerde::instance());
+    // Take ownership of the rows and free the hash table. The table will not be
+    // needed for producing spill output.
+    rowsWhileReadingSpill_ = table_->moveRows();
+    table_.reset();
+    outputPartition_ = 0;
+    nonSpilledRows_ = spiller_->finishSpill();
+  }
+
+  if (nonSpilledIndex_ < nonSpilledRows_.value().size()) {
+    uint64_t bytes = 0;
+    vector_size_t numGroups = 0;
+    auto limit = std::min<size_t>(
+        1000, nonSpilledRows_.value().size() - nonSpilledIndex_);

Please, comment about the choice of 1000.

  _____  

In velox/exec/GroupingSet.cpp <https://github.com/facebookincubator/velox/pull/1447#discussion_r875833168> :

> +    if (!merge_) {
+      merge_ = spiller_->startMerge(outputPartition_);
+    }
+    if (!mergeNext(result)) {
+      ++outputPartition_;
+      merge_ = nullptr;
+      continue;
+    }
+    return true;
+  }
+  return false;
+}
+
+bool GroupingSet::mergeNext(const RowVectorPtr& result) {
+  constexpr int32_t kBatchBytes = 1 << 20; // 1MB
+  if (mergeSelection_.size() < 64) {

Why 1MB and 64? Please, comment to explain. It would be nice to collect all the constants to a single place and document these together.

—
Reply to this email directly, view it on GitHub <https://github.com/facebookincubator/velox/pull/1447#pullrequestreview-976848171> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPT4IWKQUTGWM4HBRBVLVKTPMDANCNFSM5TZJODPQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>

mbasmanova commented:
@oerling 

> The ‘experimental’ name is literally from Presto.

We don't need to match Presto. Prestissimo workers can do the mapping as needed.

mbasmanova commented:
@oerling Orri, there are CI failures. Would you check them out?

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -611,6 +676,89 @@ TEST_F(AggregationTest, partialAggregationMemoryLimit) {
+                        .partialAggregation({"c0", "c1"}, {"array_agg(a)"})
+                        .finalAggregation()
+                        .planNode();
+  auto pair = readCursor(params, [](Task*) {});
```
consider using AssertQueryBuilder for readability

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -611,6 +676,89 @@ TEST_F(AggregationTest, partialAggregationMemoryLimit) {
+                        .planNode();
+  auto pair = readCursor(params, [](Task*) {});
+  int32_t numRows = 0;
+  for (auto rowVector : pair.second) {
```
This loop is very hard to read. Perhaps, a simpler way would be to run the aggregation without spilling and compare the results with the run w/ spilling.

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -611,6 +676,89 @@ TEST_F(AggregationTest, partialAggregationMemoryLimit) {
+
+  params.planNode = PlanBuilder()
+                        .values(batches)
+                        .partialAggregation({"c0", "c1"}, {"array_agg(a)"})
```
It doesn't feel sufficient to test just one array_agg aggregate function. It is also not clear whether the test covers paths with nulls. Perhaps, we could extend AggregationTestBase::testAggregations to include a run with spilling.

mbasmanova reviewed (commented):
@oerling Orri, I feel that test coverage is lacking. There is only one test, which is very hard to understand, to cover a lot of complex functionality with many different code paths. How are you making sure that all code paths are covered?

mbasmanova commented on velox/functions/prestosql/aggregates/tests/AggregationTestBase.h:
```diff
@@ -105,6 +110,10 @@ class AggregationTestBase : public exec::test::OperatorTestBase {
       const std::vector<std::string>& postAggregationProjections,
       const std::vector<RowVectorPtr>& expectedResult);
+
+  // Set to true if the case does not have ebough data to spill even
```
typo: ebough -> enough

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -611,6 +676,76 @@ TEST_F(AggregationTest, partialAggregationMemoryLimit) {
+
+  auto tempDirectory = exec::test::TempDirectoryPath::create();
+  std::unordered_map<std::string, std::string> configStrings;
+  configStrings[QueryConfig::kMaxPartialAggregationMemory] = "1600000";
```
nit: I wonder if these configs can be specified on the AssertQueryBuilder. This would improve readability.

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -342,6 +345,68 @@ class AggregationTest : public OperatorTestBase {
+      std::vector<RowVectorPtr>& batches) {
+    std::vector<VectorPtr> children;
+    dictionary->setSize(count * sizeof(vector_size_t));
+    children.push_back(BaseVector::wrapInDictionary(
```
nit: there is a helper method wrapInDictionary which doesn't require to pass nullptr for nulls

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -342,6 +345,68 @@ class AggregationTest : public OperatorTestBase {
+    children.push_back(BaseVector::wrapInDictionary(
+        BufferPtr(nullptr), dictionary, count, rows->childAt(1)));
+    children.push_back(children[1]);
+    batches.push_back(std::make_shared<RowVector>(
```
nit: makeRowVector(children);

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -342,6 +345,68 @@ class AggregationTest : public OperatorTestBase {
   }
 
+  // Inserts 'key' into 'order' with random bits and a serial
+  // umber. The serial number makes repeats of 'key' unique and the
```
typo: umber -> number

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -611,6 +676,76 @@ TEST_F(AggregationTest, partialAggregationMemoryLimit) {
+  using core::QueryConfig;
+  constexpr int32_t kNumDistinct = 200000;
+  constexpr int64_t kMaxBytes = 24LL << 20; // 24 MB
+  filesystems::registerLocalFileSystem();
```
this probably belongs to setup method

mbasmanova reviewed (approved):
@oerling Looks good to me. Thank you for iterating on this PR. Please, check PR title and description for typos / outdated text before landing.

oerling commented:
Thank you, Masha. Addressed the comments. 

 

 

 

Shouldn't these calculations take into account activeRows_, i.g. look for space only for a subset of rows?

 

 

A: All rows will be on here, except for null keys if null keys are dropped. Since the check is anyway approximate and will be true relatively infrequently there is no need to look at activeRows.

 

- Make directories per query?

 

Prefer not, since this is an extra metadata operation and there are no directory ops in FileSystems interface. The extra metadata op counts for something in WS and S3. File names are unique via query id and serial.

 

 

Size calculation?

 

Simplified it some. It is approximate anyway and the worst that can happen is that the insert overshoots the estimates and goes into arbitration and makes somebody else spill instead. 

 

RetainedSize?

 

Using estimateFlatSize instead.

 

 

- Hash table?

 

A: Size increment api function is a good idea.

 

- Empty variable length size?

 

A: Removed since not easily defined. Assuming that new rows are average in size is good enough. They will in practice always be at or below average.

 

Test coverage?

 

A: This covers the paths. See if fits: Fits, fits after increment, increment not possible, hence spill. Once spilling, covers opening files, switching to new files, starting wih one spilled partition and ending with all spilling. When reading results, merging with different numbers of spill files and different numbers of in memory rows of the same partition.

 

The aggregates trigger spill by either growing the accumulator or adding rows.

 

Error paths are already in SpillerTest.cpp.

 

Checking that spill bytes are set to a value in range is what's possible without adding many test-only counters.
